CLI for plotting trajectories of binaries

vtommasini opened the pull request:
## Proposed CLI for plotting trajectories
### Code review checklist

- [ ] This CLI plots trajectories from an inspiral run, concatenating the partial trajectories from each segment, and plots the full trajectories.
- [ ] The output can be a .pdf, .png, or graphic window if no output file type is sepcified.
- [ ] Example of output file:
<img width="1300" alt="Screenshot 2024-05-09 at 8 17 13 AM" src="https://github.com/sxs-collaboration/spectre/assets/164100879/9472ae18-0d20-4c79-a494-9eec128bba69">

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+)
+@click.option(
+    "--force", is_flag=True, help="Force overwriting existing output file."
+)
```
Replace these two options with:
```py
@show_or_save_plot_command()
@apply_stylesheet_command()
```
imported from `spectre.Visualization.Plot`. The first will take care of saving the plot or displaying it on screen for you. The second will apply a default configuration to the plots, and also allows the user to adjust the style (try it, e.g. `-s seaborn-dark`).

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+
+
+@click.command()
+@click.argument("simulation_name", type=str)
```
Replace this option with:

```py
@click.argument(
    "h5_files",
    nargs=-1,
    type=click.Path(exists=True, file_okay=True, dir_okay=False, readable=True),
)
```

This means the user won't pass a simulation directory to the CLI, but they will pass the reduction files directly as a glob pattern like this:

```
spectre plot trajectories path/to/Segment_*/BbhReductions.h5
```

This is more flexible because no assumptions are made on the directory structure and the user can choose which segments/files to plot. It's listed in the guidelines here: https://spectre-code.org/spectre_writing_python_bindings.html#autotoc_md155

Then you don't need the `select_reduction_files` function anymore.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+                            ][:]
+                        )
+                    )
+        except Exception as e:
+            click.echo(f"Error reading file '{filename}': {e}")
+            return None, None
```
You don't need this `try: except:`. If an error occurs then it will just propagate to the user, it'll give as much information as this `click.echo`.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+def plot_trajectory(AhA, AhB, plotlabel="", output=None, force=False):
+    indexs = [4, 5]  # 4 -> x, 5 -> y
+
+    if output and os.path.exists(output) and not force:
+        click.echo(
+            f"Warning: Output file '{output}' already exists. "
+            "Use --force to overwrite."
+        )
+        return
```
You can delete this when you use `@show_or_save_plot_command`. It's a good idea though, and we should consider adding this functionality to `@show_or_save_plot_command` in a different PR.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+        )
+        return
+
+    plt.figure(figsize=(10, 5))
```
Take the `figsize` as an additional option to the CLI command

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+    plt.figure(figsize=(10, 5))
+    plt.subplot(1, 2, 1)
+    plt.plot(
+        [AhA[i][indexs[0]] for i in range(len(AhA))],
```
Use Numpy array slicing (always prefer that over loops, it's so much faster):

`[AhA[i][indexs[0]] for i in range(len(AhA))]` -> `AhA[:, indexs[0]]`

same in multiple places below

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+        [AhA[i][indexs[0]] for i in range(len(AhA))],
+        [AhA[i][indexs[1]] for i in range(len(AhA))],
+        label="AhA",
+        color="blue",
```
`color="C0"` (that's the default and can be configured by the user)

Suggestion: I think it would be useful to fade the color from opacity 0 to 1 based on the time parameter, or maybe from some opacity < 1 to 1. That would give a sense of the time evolution. Maybe also plot `marker="."` at the very end of the trajectory to highlight the current position.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+        [AhB[i][indexs[0]] for i in range(len(AhB))],
+        [AhB[i][indexs[1]] for i in range(len(AhB))],
+        label="AhB",
+        color="red",
```
`color="C1"`

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+
+
+def plot_trajectory(AhA, AhB, plotlabel="", output=None, force=False):
+    indexs = [4, 5]  # 4 -> x, 5 -> y
```
What do you think about adding another row of plots that show the x-z projection?

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+    plt.ylabel("y1-y2")
+    plt.title("Coordinate separation" + " " + plotlabel)
+
+    plt.tight_layout()
```
You can delete this line, it's taken care of by the default plot configuration.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+
+    plt.tight_layout()
+
+    if output:
+        _, extension = os.path.splitext(output)
+        if extension.lower() not in [".pdf", ".png"]:
+            click.echo(
+                "Error: Output format not supported. "
+                "Please choose '.pdf' or '.png'."
+            )
+            return
+
+        if os.path.exists(output) and not force:
+            click.echo(
+                f"Error: Output file '{output}' already exists. "
+                "Use --force to overwrite."
+            )
+            return
+        plt.savefig(output)
+        click.echo(f"Plot saved as {output}")
+        plt.close()
+    else:
+        plt.show()  # No output file specified, display in graphic window
```
This is taken care of by `@show_or_save_plot_command`, so you can delete this part :)

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+import numpy as np
+
+
+@click.command()
```
add `name="trajectories"` so it appears properly in help texts.

Also name the function `plot_trajectories_command` for consistency.

nilsvu reviewed (changes requested):
This looks great! Thanks for writing this script. Please squash all your commits including changes you make in response to my comments below into a single commit, then force-push to this branch.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,155 @@
+    for filename in filenames:
+        try:
+            with h5py.File(filename, "r") as file:
+                if "ApparentHorizons/ControlSystemAhA_Centers.dat" in file:
```
Take the subfile names as options to the CLI command that have this as the default value.

nilsvu commented on src/Visualization/Python/PlotSizeControl.py:
```diff
@@ -27,15 +27,6 @@
     type=click.Path(exists=True, file_okay=True, dir_okay=False, readable=True),
     nargs=-1,
 )
-@click.option(
```
Oops don't delete this :)

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,157 @@
+    if AhA is not None and AhB is not None:
+        plot_trajectory(AhA, AhB)
+    else:
+        click.echo("Error: Unable to import data or not valid data.")
```
You don't need this `if: else:`. If an error happens then it will be displayed.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,159 @@
+
+    for filename in filenames:
+        with h5py.File(filename, "r") as file:
+            if subfile_name_aha in file:
```
Remove these `if` conditions as well. It's great that you're considering what happens when errors happen! I agree that's very important. You always want to give the user enough information and guidance. In this case just access the subfile so an error is thrown when it doesn't exists. That should be enough, and it's better than silently ignoring the file.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,159 @@
+            if subfile_name_ahb in file:
+                B_data.append(np.array(file[subfile_name_ahb][:]))
+
+    if A_data and B_data:
```
You can remove this `if` as well

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,159 @@
+    for filename in filenames:
+        with h5py.File(filename, "r") as file:
+            if subfile_name_aha in file:
+                A_data.append(np.array(file[subfile_name_aha][:]))
```
I suggest you import only the `time` and `xyz` columns of the data here, then you don't have to do any special indexing later. Like this: `file[subfile_name_aha][:, [0,4,5,6]]`

vtommasini committed:
Add plot trajectories CLI and unit test

fixup for PlotTrajectories CLI

fixup for Test of PlotTrajectories CLI

fixup 2 for PlotTrajectories CLI

fixup 3

fixup 4

knelli2 commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,158 @@
+)
+@show_or_save_plot_command()
+@apply_stylesheet_command()
+def plot_trajectories_command(h5_files, subfile_name_aha, subfile_name_ahb):
```
This looks great! The image that this produces is really cool!

![trajectories](https://github.com/sxs-collaboration/spectre/assets/31455476/b5b1a262-da1d-4a3a-a932-482d500263c8)

Could you also add some tests for this? Similar to the test in `tests/Unit/Visualization/Python/Test_PlotSizeControl.py`

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,158 @@
+        AhA_sampled[:, indexs[0]],
+        AhA_sampled[:, indexs[1]],
+        AhA_sampled[:, indexs[2]],
+        color="blue",
```
`color="C0"`

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,158 @@
+        return None, None
+
+
+def plot_trajectory(AhA, AhB, sample_rate=15):
```
Please add documentation to this function that describes what the function does and also describes each argument. You find examples in the other Python files in this directory.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,158 @@
+    ax1.legend()
+
+    # Calculate coordinate separation in 3D
+    x_diff_3d = AhA_sampled[:, indexs[0]] - AhB_sampled[:, indexs[0]]
+    y_diff_3d = AhA_sampled[:, indexs[1]] - AhB_sampled[:, indexs[1]]
+    z_diff_3d = AhA_sampled[:, indexs[2]] - AhB_sampled[:, indexs[2]]
```
This is just:

```
separation_3d = AhA_sampled[:, 1:] - AhB_sampled[:, 1:]
```

This computes all 3 components at once. For 2d it's `[:, 1:3]`.

vtommasini commented on src/Visualization/Python/PlotTrajectories.py:
Here is my revised version, with all the current comments/suggestions incorporated. I made the code more compact. The suggested as well as new options are documented in the help file. I'll keep working on the Test script while I wait for more feedback.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,188 @@
+        (figsize[0] * figsize[1]) / (base_figsize[0] * base_figsize[1])
+    )
+    base_fontsize = 10  # Base font size for labels
+    fontsize = base_fontsize * scale_factor
```
Is your intention here to make the font bigger when the figure is bigger? You don't need to do that. It's better to stick with the default and let the user increase the fontsize if they want (e.g. with a stylesheet).

Generally when plotting: If you want bigger labels, decrease the figsize (for reference, a paper typically has 3.4 inch wide columns). If you want higher resolution, increase the DPI (in a stylesheet).

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,188 @@
+    # Plot 3D trajectories
+    ax1 = fig.add_subplot(2, 2, 1, projection="3d")
+    ax1.plot(
+        *AhA_sampled[:, 0:].T,
```
(Optional) You can remove this indexing because it selects the full array (delete `[:, 0:]`).

nilsvu reviewed (changes requested):
Only a test is missing now as @knelli2 mentioned above. Maybe write some made-up data to a file in a `setUp` function, then run the CLI command to plot it (@knelli2 pointed to some examples). You can't really test that the plot looks correct, but at least this test makes sure that the code runs without error, and keeps working in the future.

You can squash everything into 1 commit.

nilsvu commented on tests/Unit/Visualization/Python/Test_PlotTrajectories.py:
```diff
@@ -0,0 +1,143 @@
+            unit_test_build_path(), "Visualization", "PlotTrajectories"
+        )
+        self.h5_filename = os.path.join(
+            unit_test_src_path(),
```
Put this into the `self.test_dir` as well.

More info: The `unit_test_src_path` points to the source code directory. You don't want to create temporary files there, since it can mess with git. It's better to create the data file in the test dir, which gets deleted and recreated every time you run the test.

nilsvu commented on tests/Unit/Visualization/Python/Test_PlotTrajectories.py:
```diff
@@ -0,0 +1,143 @@
+        self.create_h5_file_if_not_exists()
+
+    def tearDown(self):
+        if TestPlotTrajectories.delete_test_dir:
```
You don't need this `if`. The temporary test directory can always be deleted here.

nilsvu commented on tests/Unit/Visualization/Python/Test_PlotTrajectories.py:
```diff
@@ -0,0 +1,143 @@
+    # if h5 test file does not exist then create one using mock-up inspiral data
+    def create_h5_file_if_not_exists(self):
+        """Create the H5 file if it does not exist."""
+        if not os.path.exists(self.h5_filename):
```
You can delete this `if`. Instead, put `shutil.rmtree(self.work_dir, ignore_errors=True)` just before `os.makedirs` in the `setUp` function above. Then the test always starts from a clean slate.

nilsvu commented on tests/Unit/Visualization/Python/Test_PlotTrajectories.py:
```diff
@@ -0,0 +1,143 @@
+    def create_h5_file_if_not_exists(self):
+        """Create the H5 file if it does not exist."""
+        if not os.path.exists(self.h5_filename):
+            print(f"Creating HDF5 file: {self.h5_filename}")
```
Switch all `print`s to `logging.info()` (or delete them)

nilsvu commented on tests/Unit/Visualization/Python/Test_PlotTrajectories.py:
```diff
@@ -0,0 +1,143 @@
+        """Create the H5 file if it does not exist."""
+        if not os.path.exists(self.h5_filename):
+            print(f"Creating HDF5 file: {self.h5_filename}")
+            os.makedirs(os.path.dirname(self.h5_filename), exist_ok=True)
```
You can delete this line when the H5 file is put into the test dir.

nilsvu commented on tests/Unit/Visualization/Python/Test_PlotTrajectories.py:
```diff
@@ -0,0 +1,143 @@
+
+
+if __name__ == "__main__":
+    TestPlotTrajectories.delete_test_dir = True
```
You can delete this line

nilsvu commented on tests/Unit/Visualization/Python/Test_PlotTrajectories.py:
```diff
@@ -0,0 +1,143 @@
+            os.makedirs(os.path.dirname(self.h5_filename), exist_ok=True)
+
+            # Generate mock-up inspiral data
+            nsamples = 25000
```
This might be a bit too much data for a quick test. Reduce to just 100 points?

nilsvu commented on tests/Unit/Visualization/Python/Test_PlotTrajectories.py:
```diff
@@ -0,0 +1,143 @@
+from spectre.Visualization.PlotTrajectories import plot_trajectories_command
+
+
+class TestPlotTrajectories(unittest.TestCase):
```
Add this test to the `CMakeLists.txt` file in this directory, so it runs automatically on GitHub.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,179 @@
+    1st row: A and B trajectories in 3D, coordinate separation in 3D
+    2nd row: A and B trajectories in 2D xy plane, coordinate separation in 2D
+    xy plane. the 3D plots use a subsampled version of the data, with default
+    value of 15, to speed up the plots and avoid memory error
```
Add a description of the expected arguments, like this:

```
Arguments:
  AhA: Array of shape (num_points, 3) with the coordinates of the first object.
  AhB: ...
  sample_rate: ...
  figsize: ...
```

This is important so people know how to use your function.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,179 @@
+)
+@apply_stylesheet_command()
+@show_or_save_plot_command()
+def plot_trajectories_command(
```
Move the command below the two functions, just above `main`. That's usually the structure of a Python file so the functions are defined by the time they are used.

nilsvu commented on src/Visualization/Python/PlotTrajectories.py:
```diff
@@ -0,0 +1,179 @@
+)
+@click.option(
+    "--sample-rate",
+    "-sr",
```
We usually have single-character shortcuts. Maybe just delete this shortcut.

nilsvu commented on tests/Unit/Visualization/Python/Test_PlotTrajectories.py:
```diff
@@ -12,115 +14,110 @@
 from spectre.Visualization.PlotTrajectories import plot_trajectories_command
 
+# Configure logging
+logging.basicConfig(level=logging.INFO)
```
Move this to `__main__` at the bottom, like this:

```py
if __name__ == "__main__":
    logging.basicConfig(level=logging.DEBUG)
    unittest.main(verbosity=2)
```

nilsvu reviewed (changes requested):
Looks great! Move this one thing, then squash everything into a single commit.

nilsvu commented:
Yay nice job 👍
